openhands/llm/retry_mixin.py

```diff
@@ -0,0 +1,85 @@
+from tenacity import (
+    retry,
+    retry_if_exception_type,
+    stop_after_attempt,
+    wait_exponential,
+)
+
+from openhands.core.exceptions import OperationCancelled
+from openhands.core.logger import openhands_logger as logger
+from openhands.runtime.utils.shutdown_listener import should_exit
+
+
+class RetryMixin:
+    """Mixin class for retry logic."""
+
+    def retry_decorator(self, **kwargs):
+        """
+        Create a retry decorator with customizable parameters.
+
+        Args:
+            **kwargs: Keyword arguments to override default retry behavior.
+                      Keys: num_retries, retry_exceptions, retry_min_wait, retry_max_wait, retry_multiplier
+
+        Returns:
+            A retry decorator with the specified or default parameters.
+        """
+        num_retries = kwargs.get('num_retries')
+        retry_exceptions = kwargs.get('retry_exceptions')
+        retry_min_wait = kwargs.get('retry_min_wait')
+        retry_max_wait = kwargs.get('retry_max_wait')
+        retry_multiplier = kwargs.get('retry_multiplier')
+
+        def custom_completion_wait(self, retry_state):
+            """Custom wait for completion."""
+            exception = retry_state.outcome.exception() if retry_state.outcome else None
+            if exception is None:
+                return 0
+
+            # retry_min_wait = config.retry_min_wait
+            # retry_max_wait = config.retry_max_wait
+
+            # rate limit errors
+            exception_type = type(exception).__name__
+            logger.error(rf'\exception_type: {exception_type}\n')
+
+            if exception_type == 'RateLimitError':
+                retry_min_wait = 60
+                retry_max_wait = 240
```

Comment:
These need to be customizable, in my opinion. Whatever the initial value is, the user needs control over them, because the providers' values are wildly different, so the needs of the users are wildly different. I am fairly sure we also have documentation (or had) about these values, and are/were telling users in this repo issues and in docs that they can customize them.

Even within the same provider, they have tiers, and being in tier 1 and tier 5 is very different on exactly this.

Reply:
All I can say about the appropriate initial value is that it should be less than 60 seconds, because requests per minute is fairly widespread, and the minute is counted with the first request within the last 60 seconds. That means the remaining time must be less, possibly much less. We might have _just_ hit it in second 59! 😅 OK, I don't know that, but I suggest we should start waiting less, and let exponential do its thing.

Reply:
Yes, we should add extra options for rate limits. At the time I ran into these limits I only remember the lowest shown was like 90 seconds and 1 minute seemed reasonable to me because I didn't consider it is for "time from first fail".
Maybe lower the 60 to 30 till we have the extra options, that can than be put in this place? 🤔

Reply:
I suggest we do the other way around: configurable retries values are mainly for rate limits, these are the only ones that the user can have some idea how often (or not) they happen, for a particular provider. The other exceptions being retried are much more random, they happen or not, they get fixed next try or not.

So IMHO the values in config are for this, while for the others we might as well use the same, until we have a good reason to give _them_ the option to be configured separately.

Reply:
Hmm sorry to disagree on this, but you interprete the retry values now primarily for rate limits, but that's not how they are described by default.
If we add new options, wouldn't their name include a hint about rate limits? 🤔

Reply:
They were described at some point, as for 429 / rate limits, and other exceptions. I don't remember where or if it got lost somehow. In `llm_config.py` comments it still says something like "For models with very low limits, this can be set to 15-20." It's rather unclear, for sure, but are there other limits known in advance?

More importantly, about every topic posted in this repo about retries is about rate limits. That's what bothers people, and that can be _under their control_. People don't typically ask about retries for some random timeout, although random timeouts can happen, but their limits are probably still rate limits. 🤔 

In any case, why and how would the user customize a random, transitory exception? Why would we prioritize the customization of random exceptions over rate limiting? We can add another bunch in the poor config (3 or 4 more?), but it's not clear to me it helps a lot, and might actually confuse. Maybe we need better documentation...

Reply:
If we can use one set of retry values for all exceptions, sure, then revert the `wait=custom_completion_wait` back to the original exponential setup.
At the time, it didn't seem to me that way, but I probably made things more complicated and/worse with this. 😬

Reply:
Thanks for the discussion, it helped a lot to think it over and clarify my thoughts! And I think it's fine to use our customization options for rate limits, but we really need docs. 😅

Reply:
Documentation is like "Beetlejuice!" for developers 🤣